### ISSUE TITLE
Updating the element under specified label. Possible?

### ISSUE THREAD

**Issue by realsergii** [nmslib/hnswlib#160]
What should happen if I add 2 different elements but with the same label?
Like this:
a = np.zeros(shape=(1,512))
b = np.ones(shape=(1,512))
p.add_items(a, 1)
p.add_items(b, 1)

I thought it would be cool to update in this way the element under specified index.

If I get it right, currently I should do the following to achieve my expected result:
p.add_items(a, 1)
p.mark_deleted(1)
p.add_items(b, 1)
but the element stays in memory.

**Comment by realsergii** [comment]
After Yury's comment here https://github.com/nmslib/hnswlib/issues/72#issuecomment-549405512 I updated the code (wrapped all indexes into arrays) and double checked: what I want to achieve seems working (though updating elements in HNSW index is not officially supported, including Faiss). So I just want to make sure it's actually working as I want and expect.

import hnswlib
import numpy as np
p = hnswlib.Index(space='l2', dim=512)
p.init_index(max_elements = 10000, ef_construction = 200, M = 80)
a = np.zeros(shape=(1,512))
b = np.ones(shape=(1,512))
p.set_ef(200)
p.add_items(a, [1])
p.add_items(b, [1])
p.get_items([1])
[[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,......

**Comment by yurymalkov** [comment]
@realsergii 
Yes, it only marks the element as deleted, but it still stays in the memory. The search procedure takes it into account.

Making a full support of deletion is somewhat harder (especially if you want to do it thread safe) and not implemented here yet.

**Comment by realsergii** [comment]
@yurymalkov thanks, but this question is about updating, not deletion. So is updating possible?

**Comment by yurymalkov** [comment]
@realsergii 
It depends on strength of the update. If it is a minor (in a sense that the ordering of the neighbor does not changes compeletely) it is easy to implement. I think now you can do it via the C++ api.

If updates are big, they are equivalent to deleting and adding elements.

**Comment by realsergii** [comment]
@yurymalkov ok, but what I'm doing in my snippet (adding two different elements, but with the same label)
p.add_items(a, [1])
p.add_items(b, [1])
is it equivalent to updating an element under the label 1?

**Comment by yurymalkov** [comment]
@realsergii 
Yes, it is designed to work that way.
I suspect there might be a bug with its implementation in the master though. I'll check it.

**Comment by yurymalkov** [comment]
@realsergii There was a logical bug there.
Now it is fixed in `develop`

**Comment by realsergii** [comment]
@yurymalkov thanks! when to expect in `master` ?

**Comment by yurymalkov** [comment]
@realsergii A few days, I think

**Comment by rozendetr** [comment]
Hello. However, this dont work, when you use space='cosine'. I can not understand this. After update, i get list of [...0.04419417306780815...]

**Comment by yurymalkov** [comment]
Hi, you mean that it does not work for `cosine` in any way?
